### PR TITLE
Rank features groups obs

### LINKED ISSUE #593: rank feature groups should also be possible to be applied to obs
### Description of feature

It has been requested by multiple people that rank feature group should somehow also work for obs.

- [ ] Investigate whether and how it's possible to do so. As few API changes as possible
- [ ] Implement it
- [ ] Add tests

### PR DESCRIPTION
**Pull request by eroell**
**PR Checklist**

-   [x] This comment contains a description of changes (with reason)
-   [x] Referenced issue is linked
-   [x] If you've fixed a bug or added code that should be tested, add tests!
-   [x] Documentation in `docs` is updated (via docstrings)

**Description of changes**
Resolves #593.
As outlined in this issue, this PR adds an additional functionality for the user to use `rank_features_groups`: for the `.obs `field only (as opposed to using `.X`/`.layers`).
It does so by allowing the user to specify an additional argument `rank_obs_columns` in `rank_features_groups`. It is by default `None`, and takes the values `"all"` or a list of keys if only a selection of observation columns is wished/meaningful to be ranked.

**Technical details**
Uses `move_to_x` and `encode` (`autodetect=True`, `encodings="label"`). Makes no changes to the passed adata except for adding the respective entries in `adata.uns`.

**Questions**
- What do you think in general?
- Remarks on code?
- What do you think of this new option for the user? More arguments? Better argument name suggestions?
- Should we we want to take care of strange stuff (e.g. dates/times) or expect user to only specify columns that make sense & have it blow up if this is not done properly
- More tests?

**Additional context**
Example:

```py
import ehrapy as ep
import numpy as np
import pandas as pd
import scanpy as sc

def create_dummy_dataset_numerical_in_obs():
    """
    Create a dummy dataset with numerical and non-numerical variables in .obs.
    Also, has numerical variables in .X.
    """
    dummy_obs = {"disease":['A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'C', 'C', 'C', 'C'],
                "station": ['ICU', 'ICU', 'MICU', 'MICU', 'ICU', 'ICU', 'MICU', 'MICU', 'ICU', 'ICU', 'MICU', 'MICU'],
                "syst_bp_entry": [138, 139, 140, 141, 148, 149, 150, 151, 158, 159, 160, 161],
                "diast_bp_entry": [78, 79, 80, 81, 77, 78, 79, 80, 55, 56, 57, 58]}

    dummy_var = pd.DataFrame({"Unit": ["mg/dl", "kg"]})
    dummy_var.index = ['glucose', "weight"]
    dummy_X = np.array([[80, 90, 120, 130, 80, 130, 120, 90, 80, 90, 120, 130],
                        [77, 76, 60, 90, 110, 78, 56, 76, 67, 82, 59, 81]]).T

    adata_dummy = sc.AnnData(X=dummy_X, obs=dummy_obs, var=dummy_var)

    adata_dummy.uns['numerical_columns'] = ['glucose', 'weight']
    adata_dummy.uns['non_numerical_columns'] = []
    adata_dummy.uns['encoded_non_numerical_columns'] = []
    
    return adata_dummy

adata_dummy = create_dummy_dataset_numerical_in_obs()

# classical way of using .X
ep.tl.rank_features_groups(adata_dummy, groupby='disease',)
ep.pl.rank_features_groups(adata_dummy)
```
image

```py
# added way of using .obs
ep.tl.rank_features_groups(adata_dummy, groupby='disease', rank_obs_columns="all", key_added="rank_obs")
ep.pl.rank_features_groups(adata_dummy, key="rank_obs")
```
image

### CODE REVIEW

**Comment by juzb** [comment]
@eroell this is exactly what I had in mind! Really like the None / all / some_list options too :)

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 1]
```diff
@@ -255,6 +257,7 @@ def rank_features_groups(
     correction_method: _method_options._correction_method = "benjamini-hochberg",
     tie_correct: bool = False,
     layer: Optional[str] = None,
+    rank_obs_columns: Optional[Union[list[str], str]] = None,
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:260]: 1. Can we pls import `from __future__ import __annotations__`?
2. We can probably use a different type then `list`. `Iterable` or `Sequence`?

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 2]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
          >>> ep.tl.rank_features_groups(adata, "service_unit")
          >>> ep.pl.rank_features_groups(adata)
     """
+    # if rank_obs_columns is indicated, layer must be None
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:328]:
```suggestion
```

The code already says it :)

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 3]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+    if layer is not None and rank_obs_columns is not None:
+        raise ValueError("Only one of layer and rank_obs_columns can be specified.")
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:330]:
```suggestion
        raise ValueError("Only one of 'layer' and 'rank_obs_columns' can be specified.")
```

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 4]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+
     adata = adata.copy() if copy else adata
 
+    if rank_obs_columns is not None:
+        # keep reference to original adata, needed if copy=False
+        adata_orig = adata
+        # copy adata to work on
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:337]:
```suggestion
```

yup, that's what the code says

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 5]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+        adata = adata.copy()
+
+        if isinstance(rank_obs_columns, str):
+            if rank_obs_columns == "all":
+                rank_obs_columns = adata.obs.keys().to_list()
+            else:
+                raise ValueError(
+                    f"rank_obs_columns should be 'all' or Iterable of column names, not {rank_obs_columns}."
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:345]: Yeah, you said it yourself. It's an Iterable and not just a list. Let's use this for our typehint!

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 6]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+                )
+
+        # consider adata where all columns from obs become the features, and the other features are dropped
+        if not all(elem in adata.obs.columns.values for elem in rank_obs_columns):
+            raise ValueError(
+                f"Columns `{[col for col in rank_obs_columns if col not in adata.obs.columns.values]}` are not in obs."
+            )
+
+        # if groupby in rank_obs_columns:
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:354]: Can we remove the outcommented dead code, please?

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 7]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+        #     rank_obs_columns.remove(groupby)
+
+        # move obs columns to X
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:357]:
```suggestion
```
That's what the code below says. Explain WHY if necessary.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 8]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+        adata_with_moved_columns = move_to_x(adata, rank_obs_columns)
+
+        # remove columns previously in X
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:360]:
```suggestion
```
That's what the code below says. Explain WHY if necessary.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 9]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+        columns_to_select = adata_with_moved_columns.var_names.difference(adata.var_names)
+        adata_with_moved_columns = adata_with_moved_columns[:, columns_to_select]
+
+        # encode categoricals
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:364]:
```suggestion
```

That's what the code below says. Explain WHY if necessary.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 10]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+        adata_with_moved_columns = encode(adata_with_moved_columns, autodetect=True, encodings="label")
+
+        # assign numeric and categorical columns
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:367]:
```suggestion
```
That's what the code below says. Explain WHY if necessary.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 11]
```diff
@@ -320,8 +325,54 @@ def rank_features_groups(
...
+    if layer is not None and rank_obs_columns is not None:
+        raise ValueError("Only one of layer and rank_obs_columns can be specified.")
+
     adata = adata.copy() if copy else adata
 
+    if rank_obs_columns is not None:
+        # keep reference to original adata, needed if copy=False
+        adata_orig = adata
+        # copy adata to work on
+        adata = adata.copy()
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:338]: Such a copy can be very expensive. Is it possible to instead create an intermediate AnnData object with only the information that we need?

**Review by Zethson — CHANGES_REQUESTED** [review]
Awesome! A few minor comments.

Generally, I have the impression that you could leave out a few code comments. I myself comment a lot and am actually learning to reduce it a bit. Hence, I gave a few suggestions.

**Comment by Zethson** [comment]
> What do you think of this new option for the user? More arguments? Better argument name suggestions?

Perfectly fine.

> Should we we want to take care of strange stuff (e.g. dates/times) or expect user to only specify columns that make sense & have it blow up if this is not done properly

Maybe we should check the dtype of the obs columns and shout at least if we saw dates/times?


> More tests?

Only if we're handling date and time. Otherwise LGTM

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 2 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py:328]: Agreed :D

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 3 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py:329]: Agreed

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 1 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py:260]: 1. Yes see no reason against that
2. Agree, had this in the first place - then ruff then yielded an error: `move_to_x` expecting list[str] | str, but incoming `Iterable[str]`. Suppose can just convert Iterable to list in `move_to_x`

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 5 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py:343]: Yes indeed - wrote that when had Iterable first, before stumbling across move_to_x only taking list.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 6 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py]: Yep done

File: ehrapy/tools/feature_ranking/_rank_features_groups.py (hunk 11 above)
Comment (eroell) [reply ehrapy/tools/feature_ranking/_rank_features_groups.py]: Yes I think so - the latest commit only keeps the `.obs `field (and for conciseness one dummy column of zeros in `.X`)

**Comment by eroell** [comment]
New version now can consider highly variable features in `.X`/`.layers[layer]`, `.obs`, or both.

This introduces two new arguments:
- `field_to_rank`: available options are `'layer'` (default), `'obs'` (what has been possible above), `'layer_and_obs'` (new here)
- `columns_to_rank`: default `'all'`, but can be dict with `var_names` and `obs_names` keys, e.g. `{"var_names": ["glucose"], "obs_names": ["syst_bp_entry", "station"]}`. The ability to select columns from .obs is probably important due to many columns likely not of interest (such as dates); Since var and obs is less strictly separated here than in single-cell though, having the column selection option for both cases probably makes sense, especially if combining them in an analysis this way...

But it might be that this has become overengineered, with more pitfalls of how it could raise Errors in a confusing way

@Zethson what do you think?
@juzb once more - what do you think of the examples below? More confusing to use than the example above? 

**Example**

```py
import ehrapy as ep
import numpy as np
import pandas as pd
import scanpy as sc

def create_dummy_dataset_numerical_in_obs():
    """
    Create a dummy dataset with numerical and non-numerical variables in obs.
    Also, has numerical variables in .X.
    """
    dummy_obs = {"disease":['A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'C', 'C', 'C', 'C'],
                "station": ['ICU', 'ICU', 'MICU', 'MICU', 'ICU', 'ICU', 'MICU', 'MICU', 'ICU', 'ICU', 'MICU', 'MICU'],
                "syst_bp_entry": [138, 139, 140, 141, 148, 149, 150, 151, 158, 159, 160, 161],
                "diast_bp_entry": [78, 79, 80, 81, 77, 78, 79, 80, 55, 56, 57, 58]}

    dummy_var = pd.DataFrame({"Unit": ["mg/dl", "kg"]})
    dummy_var.index = ['glucose', "weight"]
    dummy_X = np.array([[80, 90, 120, 130, 80, 130, 120, 90, 80, 90, 120, 130],
                        [77, 76, 60, 90, 110, 78, 56, 76, 67, 82, 59, 81]]).T

    adata_dummy = sc.AnnData(X=dummy_X, obs=dummy_obs, var=dummy_var)

    adata_dummy.uns['numerical_columns'] = ['glucose', 'weight']
    adata_dummy.uns['non_numerical_columns'] = []
    adata_dummy.uns['encoded_non_numerical_columns'] = []
    
    return adata_dummy

# classical way of using .X
adata_dummy = create_dummy_dataset_numerical_in_obs()
ep.tl.rank_features_groups(adata_dummy, groupby='disease')
ep.pl.rank_features_groups(adata_dummy)
```
image

```py
# added way of using .obs
adata_dummy = create_dummy_dataset_numerical_in_obs()
ep.tl.rank_features_groups(adata_dummy, groupby='disease', field_to_rank="obs")
ep.pl.rank_features_groups(adata_dummy)
```
image

```py
# added way of using both
adata_dummy = create_dummy_dataset_numerical_in_obs()
ep.tl.rank_features_groups(adata_dummy, groupby='disease', field_to_rank="layer_and_obs")
ep.pl.rank_features_groups(adata_dummy)
```
image

```py
# added way of using e.g. both, with selecting only a some variables
adata_dummy = create_dummy_dataset_numerical_in_obs()
ep.tl.rank_features_groups(adata_dummy, groupby='disease',
    field_to_rank="layer_and_obs",
    columns_to_rank={"var_names": ["glucose"],
                                   "obs_names": ["syst_bp_entry", "station"]
                                   }
)
```
image

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 12]
```diff
@@ -240,6 +244,51 @@ def _evaluate_categorical_features(
     )
 
 
+def _check_no_datetime_columns(df):
+    datetime_cols = [col for col in df.columns if df[col].dtype == "datetime64[ns]"]
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:248]:
```suggestion
datetime_cols = [col for col in df.columns if pd.api.types.is_datetime64_any_dtype(df[col]) or pd.api.types.is_timedelta64_dtype(df[col])]
```

Just an idea. Not sure whether it's needed or better.

File: ehrapy/tools/feature_ranking/_rank_features_groups.py [hunk 13]
```diff
@@ -320,8 +373,88 @@ def rank_features_groups(
...
+            X_to_keep = adata.X if layer is None else adata.layers[layer]
+            var_to_keep = adata.var
+            uns_num_to_keep = adata.uns["numerical_columns"] if "numerical_columns" in adata.uns else []
+            uns_enc_to_keep = (
+                adata.uns["encoded_non_numerical_columns"] if "encoded_non_numerical_columns" in adata.uns else []
+            )
+            uns_non_num_to_keep = adata.uns["non_numerical_columns"] if "non_numerical_columns" in adata.uns else []
+
+    else:
+        X_to_keep = np.zeros((len(adata), 1))
```
Comment (Zethson) [inline ehrapy/tools/feature_ranking/_rank_features_groups.py:418]: This is a section where you could introduce some reasoning. Why the dummy values? I'd add a short comment above this section

**Review by Zethson — APPROVED** [review]
Great! Let's ignore ruff for now. Maybe it sorts itself out. If not, I'll have a proper look

**Pull request merged**